go/vt/vtgate/planbuilder/operators/route.go

```diff
@@ -672,23 +672,81 @@ func addMultipleColumnsToInput(
 		return "", op, true, offset
 
 	case *Union:
-		tableID := semantics.SingleTableSet(len(ctx.SemTable.Tables))
-		ctx.SemTable.Tables = append(ctx.SemTable.Tables, nil)
-		unionColumns := op.GetColumns(ctx)
-		proj := &Projection{
-			Source:  op,
-			Columns: AliasedProjections(slice.Map(unionColumns, newProjExpr)),
-			DT: &DerivedTable{
-				TableID: tableID,
-				Alias:   "dt",
-			},
-		}
+		proj := addDerivedProj(ctx, op)
 		return addMultipleColumnsToInput(ctx, proj, reuse, addToGroupBy, exprs)
 	default:
 		return "", op, false, nil
 	}
 }
 
+func (r *Route) AddWSColumn(ctx *plancontext.PlanningContext, offset int, _ bool) int {
+	columns := r.GetColumns(ctx)
+	if offset > len(columns) {
+		panic(vterrors.VT13001("column %d not found", offset))
+	}
+	col := columns[offset]
+	if offset := r.FindCol(ctx, weightStringFor(col.Expr), true); offset >= 0 {
+		return offset
+	}
+
+	ok, foundOffset := addWSColumnToInput(ctx, r.Source, offset)
+	if !ok {
+		src := addDerivedProj(ctx, r.Source)
+		r.Source = src
+		return src.AddWSColumn(ctx, offset, true)
+	}
+	return foundOffset
+}
+
+func addWSColumnToInput(ctx *plancontext.PlanningContext, source Operator, offset int) (bool, int) {
+	switch op := source.(type) {
+	case *SubQuery:
+		return addWSColumnToInput(ctx, op.Outer, offset)
+	case *Distinct:
+		return addWSColumnToInput(ctx, op.Source, offset)
+	case *Filter:
+		return addWSColumnToInput(ctx, op.Source, offset)
+	case *Projection:
+		return true, op.AddWSColumn(ctx, offset, true)
+	case *Aggregator:
+		return true, op.AddWSColumn(ctx, offset, true)
+	}
+	return false, -1
+}
+
+func addDerivedProj(
+	ctx *plancontext.PlanningContext,
+	op Operator,
+) (projection *Projection) {
+	// create new table ID
+	tableID := semantics.SingleTableSet(len(ctx.SemTable.Tables))
+	ctx.SemTable.Tables = append(ctx.SemTable.Tables, nil)
```

Comment:
nit: more comments can be added here as adding `nil` to semtable tables is not clear. I understand the intention here but comments will help.

Reply:
Ideally we can add the `DerivedTable` here which is created below.

Reply:
the Derived table that we create here is of a very different type than what we store in the semantic state, and the info that we normally store there does not really make sense to store here. we just need an ID